### PR TITLE
[`pathbuf_init_then_push`]:  Checks for calls to `push` immediately a…

### PR DESCRIPTION
**Pull request by lengyijun**
changelog: [`pathbuf_init_then_push`]: new lint: Checks for calls to `push` immediately after creating a new `PathBuf`

Just a mirror of VEC_INIT_THEN_PUSH

### CODE REVIEW

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (Centri3) [inline clippy_lints/src/pathbuf_init_then_push.rs:27]: Why wouldn't you use `PathBuf::from` here? We should lint both of these and suggest it instead, imo.

I suppose this would insert `/` or `\\` automatically but that's all I can really think of.

File: CHANGELOG.md
Comment (Centri3) [inline CHANGELOG.md:5344]: Please run `cargo dev update_lints`

**Comment by lengyijun** [comment]
Someone may care about the performance: .join() introduce an implicit clone
But personally, I don't care

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (lengyijun) [reply clippy_lints/src/pathbuf_init_then_push.rs:27]: I think this is very rare
Moreover, I'm not familar with the details of pathbuf

Comment (lengyijun) [reply clippy_lints/src/pathbuf_init_then_push.rs:30]: I doubt it's not trial after reading https://github.com/rust-lang/rust-clippy/pull/11453

Comment (Centri3) [reply clippy_lints/src/pathbuf_init_then_push.rs:30]: wdym? I'm not sure why `PathBuf::new().join("foo").join("bar")` can't be `PathBuf::from("foo/bar")` (`/` is crossplatform iirc)

Comment (lengyijun) [reply clippy_lints/src/pathbuf_init_then_push.rs:30]: I don't like overeager suggestion

**Comment by xFrednet** [comment]
Hey @lengyijun, would you mind rebasing this on master?

@ARandomDev99 This might be a good PR to review. There are some conflicts, but the implementation should be unaffected by the rebase.

If you comment something on this PR, I can also add you as an assignee

r? xFrednet

**Comment by ARandomDev99** [comment]
I concur with Centri3. Instead of multiple calls to `PathBuf::join`, it is better to suggest using `PathBuf::from` while dealing with only string literals. Additionally, I believe this lint should be in the `pedantic` category instead of `complexity`.

**Comment by xFrednet** [comment]
>  Additionally, I believe this lint should be in the `pedantic` category instead of `complexity`.

Could you explain why you feel this should be `pedantic`?

`clippy::vec_init_then_push` is warn-by-default and if this is free from FPs I think it's reasonable to have it as warn-by-default as well. We can bikeshed this part and leave it to the FCP before we merge this :)

**Comment by lengyijun** [comment]
blocked by https://github.com/rust-lang/rust/issues/123352

**Comment by ARandomDev99** [comment]
> Could you explain why you feel this should be `pedantic`?

@lengyijun said:
> Someone may care about the performance: .join() introduce an implicit clone

This doesn't seem like a change that everyone would prefer. Having it allowed by default seems more appropriate to me. Unlike `clippy::vec_init_then_push`, which improves performance, this lint only suggests replacing calls to `push` with `join`.

**Comment by xFrednet** [comment]
> This doesn't seem like a change that everyone would prefer.

Ahh, yep you're right. Then `pedantic` would be the safe choice :+1: 

> I'm still relatively new around here. What does FCP mean? 😅

No problem, FCP stands for *final comment period* it's a new concept for Clippy. When a new lint is ready to be merged, we create a topic on zulip to get feedback. If no concerns are raised within ~1 week we merge the PR, otherwise we discuss the issues :)

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (xFrednet) [inline clippy_lints/src/pathbuf_init_then_push.rs:86]: I think here it would be good to also check for `#[cfgs]` as the `push` might be conditional: 

```suggestion
    fn display_err(&self, cx: &LateContext<'_>) {
        if clippy_utils::attrs::span_contains_cfg(self.err_span) { return; }
```

File: tests/ui/pathbuf_init_then_push.rs
Comment (xFrednet) [inline tests/ui/pathbuf_init_then_push.rs:1]: Could you also add some tests for code snippets that shouldn't trigger the lint. Something like:

```rs
let mut x = PathBuf::new();
println!("{}", x.display());
x.push("Duck");
```

And also this test, which should trigger the lint:

```rs
let mut path_buf = PathBuf::from("foo").join("bar");
path_buf.push("buz");
```

**Review by xFrednet — COMMENTED** [review]
The implementation looks good to me overall. I would like to see more tests, also negative examples of code that shouldn't trigger the lint. I left some example comments. 

There is also a random binary file called `c`. Can you remove that one?

If you rebase, we could also take a look at the lintcheck output in our CI :D

----

Sorry, that it took this long, it again slipped my mind :see_no_evil:

**Comment by xFrednet** [comment]
Let me know when this PR is ready to be reviewed again. I'm assuming the last commits are not ready yet due to their name :)

File: tests/ui/pathbuf_init_then_push.rs
Comment (lengyijun) [reply tests/ui/pathbuf_init_then_push.rs]: no interest

Comment (lengyijun) [reply tests/ui/pathbuf_init_then_push.rs]: no interest

Comment (lengyijun) [reply tests/ui/pathbuf_init_then_push.rs]: ready to review

Comment (xFrednet) [reply tests/ui/pathbuf_init_then_push.rs]: I don't understand the "no interest" comments, were they intended for me?

Could you also add the example from my comment as a test?

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (xFrednet) [reply clippy_lints/src/pathbuf_init_then_push.rs:33]: Now it would make sense to change this to `PathBuf::from` since that's an actual suggestion of the lint.

You can add two suggestions in the code description. Just have a comment like:

```
// Use PathBuf::from:
...

// You can also use `.join()`
...
```

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (xFrednet) [inline clippy_lints/src/pathbuf_init_then_push.rs:59]: NIT:
```suggestion
    /// Try to generate a suggestion with `PathBuf::from`.
    /// Returns `None` if the suggestion would be invalid. 
```

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (xFrednet) [inline clippy_lints/src/pathbuf_init_then_push.rs:150]: This should probably also clear the `searcher`. If there is an assign expression between the initial `PathBuf` creation and the `x.push(y)` call, it's better not to lint. This could for instance modify `y`. 

```suggestion
        if let Some(mut searcher) = self.searcher.take() 
```

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (xFrednet) [inline clippy_lints/src/pathbuf_init_then_push.rs:189]: The `searcher.display_err(cx)` should never be executed, since the `x.push();` statement from `check_stmt` always emits the lint if there is something to lint.

```suggestion
    fn check_block_post(&mut self, cx: &LateContext<'tcx>, _: &'tcx Block<'tcx>) {
        self.searcher = None;
    }
```

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (xFrednet) [reply clippy_lints/src/pathbuf_init_then_push.rs:150]: If the lint no longer triggers after this, it could be that the `check_expr` call from the variable `init` expression calls this function and clears the searcher.

I think the easiest solution would be to move this and the `check_local` code into `check_stmt` and handle everything from the statement level downwards

**Review by xFrednet — COMMENTED** [review]
I left some more comments, they should hopefully all be easy to fix. They are hopefully not too nitpicky ^^'

Let me know, if you have any question :D

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (xFrednet) [reply clippy_lints/src/pathbuf_init_then_push.rs:36]: @lengyijun have you seen this comment?

**Comment by lengyijun** [comment]
https://github.com/rust-lang/rust-clippy/blob/279494eced419dca11c4b8e589fc6cec9dc82f46/lintcheck/src/input.rs#L182C16-L187C19

Emmm

**Comment by xFrednet** [comment]
I assume the `Emmm` is referring to the comment?

You can use [span_contains_comment](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.span_contains_comment.html#) to check if there is a comment between the creation and the `push`

**Comment by xFrednet** [comment]
Okay, I'd say the FCP has concluded. It sounds like the lint is okay, but we had others suggest the `restriction` group instead, since it might be more readable but has a performance overhead. Could you change the lint group? Then it should be ready for merge. With the addition of [span_contains_comment](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.span_contains_comment.html#) as mentioned above :D

File: clippy_lints/src/pathbuf_init_then_push.rs
Comment (xFrednet) [inline clippy_lints/src/pathbuf_init_then_push.rs:93]: NIT: Could you add a test for this? Something like this:

```rs
   let mut path_buf = PathBuf::new();
   #[cfg(cats)]
   path_buf.push("foo");
```

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
This version looks good to me, and the two detected cases in our CI are also very good: https://github.com/rust-lang/rust-clippy/actions/runs/10004116830?pr=11700

Thank you for the last changes. I left a tiny nit, which should be simple to fix. Then you can `r=me`

---

Roses are red,
This PR is good,
I delegate this,
To bors the bot

**Pull request merged**
